pilot/pkg/xds/eds.go

```diff
@@ -551,7 +553,7 @@ func (eds *EdsGenerator) buildEndpoints(proxy *model.Proxy,
 	req *model.PushRequest,
 	w *model.WatchedResource) (model.Resources, model.XdsLogDetails) {
 	var edsUpdatedServices map[string]struct{}
-	if !req.Full {
+	if !req.Full || onlyEndpointsChanged(req) {
```

Comment:
NOTE: I built this on top of https://github.com/istio/istio/pull/37974 to make things simpler. I will back it out if we don't merge that, and put a hold on this PR unti lthat one lands